### ISSUE TITLE
Missing warning about removing `mut` qualifier?

### ISSUE THREAD

**Issue by djc** [rust-lang/rust#50321]
During work on Cargo, I found a seeming bug where the compiler doesn't warn about an unnecessary `mut` qualifier. This version does not warn:

```rust
fn bar<'a>(foo: &'a String) -> bool {
    foo == "foo"
}

fn main() {
    let res = {
        let mut foo = String::from("foo");
        bar(&foo)
    };
    println!("{:?}", res);
}
```

This version does warn:

```rust
fn bar<'a>(foo: &'a String) -> bool {
    foo == "foo"
}

fn main() {
    let mut foo = String::from("foo");
    let res = bar(&foo);
    println!("{:?}", res);
}
```

**Comment by estebank** [comment]
[It warns now](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f45b1e6a54757617d4c567b81cb72c0e):

```
warning: variable does not need to be mutable
 --> src/main.rs:8:13
  |
8 |         let mut foo = String::from("foo");
  |             ----^^^
  |             |
  |             help: remove this `mut`
  |
  = note: #[warn(unused_mut)] on by default)
```

**Comment by djc** [comment]
Yeah, I suppose this has been fixed with NLL.
